include/minisketch.h

```diff
@@ -151,6 +151,156 @@ size_t minisketch_compute_max_elements(uint32_t bits, size_t capacity, uint32_t
+    {
+        m_minisketch = std::unique_ptr<minisketch, Deleter>(minisketch_create(bits, implementation, capacity));
+    }
+
+    /** Create a Minisketch object sufficiently large for the specified number of elements at given fpbits. */
+    static Minisketch CreateFP(uint32_t bits, uint32_t implementation, size_t max_elements, uint32_t fpbits) noexcept
+    {
+        return Minisketch(bits, implementation, ComputeCapacity(bits, max_elements, fpbits));
+    }
+
+    /** See minisketch_get_bits(). */
+    uint32_t GetBits() const noexcept { return minisketch_bits(m_minisketch.get()); }
+
+    /** See minisketch_get_capacity(). */
+    size_t GetCapacity() const noexcept { return minisketch_capacity(m_minisketch.get()); }
+
+    /** See minisketch_get_implementation(). */
+    uint32_t GetImplementation() const noexcept { return minisketch_implementation(m_minisketch.get()); }
+
+    /** See minisketch_set_seed(). */
+    Minisketch& SetSeed(uint64_t seed) noexcept
+    {
+        minisketch_set_seed(m_minisketch.get(), seed);
+        return *this;
+    }
+
+    /** See miniksetch_add_element(). */
+    Minisketch& Add(uint64_t element) noexcept
+    {
+        minisketch_add_uint64(m_minisketch.get(), element);
+        return *this;
+    }
+
+    /** See minisketch_merge(). */
+    Minisketch& Merge(const Minisketch& sketch) noexcept
+    {
+        minisketch_merge(m_minisketch.get(), sketch.m_minisketch.get());
+        return *this;
+    }
+
+    /** Decode this sketch into the result vector, which must have a size equal to the maximum allowed elements. */
+    bool Decode(std::vector<uint64_t>& result) const noexcept
+    {
+        ssize_t ret = minisketch_decode(m_minisketch.get(), result.size(), result.data());
+        if (ret == -1) return false;
+        result.resize(ret);
+        return true;
+    }
+
+    /** See minisketch_serialized_size(). */
+    size_t GetSerializedSize() const noexcept { return minisketch_serialized_size(m_minisketch.get()); }
+
+    /** Serialize the sketch as a byte vector. */
+    std::vector<unsigned char> Serialize() const noexcept
+    {
+        std::vector<unsigned char> result(GetSerializedSize());
+        minisketch_serialize(m_minisketch.get(), result.data());
+        return result;
+    }
+
+    /** Deserialize into this sketch from an object containing its bytes (which has data() and size() members). */
+    template<typename T>
+    Minisketch& Deserialize(
+        const T& obj,
+        typename std::enable_if<
+            std::is_convertible<typename std::remove_pointer<decltype(obj.data())>::type (*)[], const unsigned char (*)[]>::value &&
+            std::is_convertible<decltype(obj.size()), std::size_t>::value,
+            std::nullptr_t
+        >::type = nullptr) noexcept
+    {
+        assert(GetSerializedSize() == obj.size());
+        minisketch_deserialize(m_minisketch.get(), obj.data());
+        return *this;
```

Comment:
Why does this have to return anything if it's called on a sketch instance?

Reply:
It's so that you can chain operations. E.g. `auto vec = Minisketch::CreateFP(...).Deserialize(v).Add(x).Add(y).Merge(z).Serialize();`

Reply:
I assume this is going to be safe even if the instance evaluates to `false` due to implementation error nullptr?

Reply:
I added a `Minisketch::SupportedImplementation()` to detect ahead of time which configurations are legal. You shouldn't call `Minisketch::CreateFP()` or the `Minisketch` constructor for unsupported combinations.